packages/flutter/lib/src/cupertino/text_field.dart

```diff
@@ -541,12 +556,36 @@ class _CupertinoTextFieldState extends State<CupertinoTextField> with AutomaticK
   }
 
   void _handleSingleLongTapEnd(LongPressEndDetails details) {
-    _editableTextKey.currentState.showToolbar();
+    if (_shouldShowSelectionOverlay)
+      _editableText.showToolbar();
   }
 
   void _handleDoubleTapDown(TapDownDetails details) {
     _renderEditable.selectWord(cause: SelectionChangedCause.tap);
-    _editableTextKey.currentState.showToolbar();
+    if (_shouldShowSelectionOverlay)
+      _editableText.showToolbar();
+  }
+
+  bool _shouldShowHandles(SelectionChangedCause cause) {
+    // When the text field is activated by something that doesn't trigger the
+    // selection overlay, we shouldn't show the handles either.
+    if (!_shouldShowSelectionOverlay)
+      return false;
+
+    // On iOS, we don't show handles when the selection is collapsed.
+    if (_effectiveController.selection.isCollapsed)
+      return false;
+
+    if (cause == SelectionChangedCause.keyboard)
+      return false;
+
+    if (cause == SelectionChangedCause.longPress)
```

Comment:
This might need a comment above. Isn't long press selections on iOS always collapsed?